docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,597 @@
+
+IMPORTANT: A build step that consumes the `ValidationPhaseBuildItem` should always produce a `ValidationErrorBuildItem` or at least inject a `BuildProducer` for this build item, otherwise it could be ignored or processed at the wrong time (e.g. after the correct CDI bootstrap phase).
+
+[source,java]
+----
+@BuildStep
+void validate(ValidationPhaseBuildItem validationPhase,
+            BuildProducer<MyBuildItem> myBuildItem, 
+            BuildProducer<ValidationErrorBuildItem> errors) {
+   if (someCondition) {
+     errors.produce(new ValidationErrorBuildItem(new IllegalStateException()));
+     myBuildItem.produce(new MyBuildItem());
+   }
+}
+----
+
+[[custom_context]]
+== Use Case - Register a Custom CDI Context
+
+Sometimes extensions need to extend the set of built-in CDI scopes and contexts.
+
+=== _Solution 1_: `ContextRegistrarBuildItem`
+
+If you need to add a custom context and you're not going to produce a build item within the registration then use the `ContextRegistrarBuildItem` and leverage the `io.quarkus.arc.processor.ContextRegistrar` API.
+
+.`ContextRegistrarBuildItem` Example
+[source,java]
+----
+@BuildStep
+ContextRegistrarBuildItem registerContext() {
+    return new ContextRegistrarBuildItem(new ContextRegistrar() {
+         public void register(RegistrationContext context) {
+            // NOTE: You can't produce a build item from within this callback!
+            context.configure(TransactionScoped.class).normal().contextClass(TransactionContext.class).done();
+         }
+    });
+}
+----
+
+=== _Solution 2_: `ContextRegistrationPhaseBuildItem`
+
+If an extension needs to produce other build items during the registration it should use the `ContextRegistrationPhaseBuildItem` instead.
+The reason is that an injected `BuildProducer` is only valid during a `@BuildStep` method invocation.
+
+IMPORTANT: A build step that consumes the `ContextRegistrationPhaseBuildItem` should always produce a `ContextConfiguratorBuildItem` or at least inject a `BuildProducer` for this build item, otherwise it could be ignored or processed at the wrong time (e.g. after the correct CDI bootstrap phase).
+
+[source,java]
+----
+@BuildStep
+void registerContext(MyConfig config, ContextRegistrationPhaseBuildItem phase,
+            BuildProducer<MyBuildItem> myBuildItem, 
+            BuildProducer<ContextConfiguratorBuildItem> contexts) {
+   if (config.isEnabled()) {
+      ContextConfigurator configurator = phase.getContext().configure(TransactionScoped.class).normal().contextClass(TransactionContext.class).done();
+      contexts.produce(new ContextConfiguratorBuildItem(configurator));
+      myBuildItem.produce(new MyBuildItem());
+   }
+}
+----
+
+=== What if I Need to Know All the Scopes Used in the Application?
+
+_Solution_: You can inject the `CustomScopeAnnotationsBuildItem` in a build step and use the convenient methods such as `CustomScopeAnnotationsBuildItem.isScopeDeclaredOn()`.
```

Comment:
Similarly to a previous question -- sounds like this shouldn't be used in some earlier phases?

Reply:
It can be used before the bootrap actually starts.